public/app/features/alerting/unified/state/actions.ts

```diff
@@ -70,62 +83,161 @@ export function fetchAllPromAndRulerRulesAction(force = false): ThunkResult<void
   };
 }
 
-export function deleteRuleAction(ruleLocation: RuleLocation): ThunkResult<void> {
-  /*
-   * fetch the rules group from backend, delete group if it is found and+
-   * reload ruler rules
-   */
-  return async (dispatch) => {
-    const { namespace, groupName, ruleSourceName, ruleHash } = ruleLocation;
-    //const group = await fetchRulerRulesGroup(ruleSourceName, namespace, groupName);
-    const groups = await fetchRulerRulesNamespace(ruleSourceName, namespace);
-    const group = groups.find((group) => group.name === groupName);
-    if (!group) {
-      throw new Error('Failed to delete rule: group not found.');
+async function findExistingRule(ruleIdentifier: RuleIdentifier): Promise<RuleWithLocation | null> {
+  if (isGrafanaRuleIdentifier(ruleIdentifier)) {
+    const namespaces = await fetchRulerRules(GRAFANA_RULES_SOURCE_NAME);
+    // find namespace and group that contains the uid for the rule
+    for (const [namespace, groups] of Object.entries(namespaces)) {
+      for (const group of groups) {
+        const rule = group.rules.find((rule: RulerGrafanaRuleDTO) => rule.grafana_alert?.uid === ruleIdentifier.uid);
+        if (rule) {
+          return {
+            group,
+            ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
+            namespace: namespace,
+            rule,
+          };
+        }
+      }
     }
-    const existingRule = group.rules.find((rule) => hashRulerRule(rule) === ruleHash);
-    if (!existingRule) {
-      throw new Error('Failed to delete rule: group not found.');
+  } else {
+    const { ruleSourceName, namespace, groupName, ruleHash } = ruleIdentifier;
+    const group = await fetchRulerRulesGroup(ruleSourceName, namespace, groupName);
+    if (group) {
+      const rule = group.rules.find((rule) => hashRulerRule(rule) === ruleHash);
+      if (rule) {
+        return {
+          group,
+          ruleSourceName,
+          namespace,
+          rule,
+        };
+      }
     }
-    // for cloud datasources, delete group if this rule is the last rule
-    if (group.rules.length === 1 && isCloudRulesSource(ruleSourceName)) {
-      await deleteRulerRulesGroup(ruleSourceName, namespace, groupName);
+  }
+  return null;
+}
+
+export const fetchExistingRuleAction = createAsyncThunk(
+  'unifiedalerting/fetchExistingRule',
+  (ruleIdentifier: RuleIdentifier): Promise<RuleWithLocation | null> =>
+    withSerializedError(findExistingRule(ruleIdentifier))
+);
+
+async function deleteRule(ruleWithLocation: RuleWithLocation): Promise<void> {
```

Comment:
nit: I think it would be good to use early exit here instead of `if`...`else` to prevent nesting. Makes it easier to read the code.

Reply:
- [ ] early exit